src/main/java/org/apache/commons/text/numbers/DoubleFormat.java

```diff
@@ -0,0 +1,736 @@
+            Objects.requireNonNull(symbols, "Decimal format symbols cannot be null");
+
+            return digits(getDigitString(symbols))
+                    .decimalSeparator(symbols.getDecimalSeparator())
+                    .groupingSeparator(symbols.getGroupingSeparator())
+                    .minusSign(symbols.getMinusSign())
+                    .exponentSeparator(symbols.getExponentSeparator())
+                    .infinity(symbols.getInfinity())
+                    .nan(symbols.getNaN());
+        }
+
+        /** Get a string containing the localized digits 0-9 for the given symbols object. The
+         * string is constructed by starting at the {@link DecimalFormatSymbols#getZeroDigit() zero digit}
+         * and adding the next 9 consecutive characters.
+         * @param symbols symbols object
+         * @return string containing the localized digits 0-9
+         */
+        private String getDigitString(final DecimalFormatSymbols symbols) {
+            final int zeroDelta = symbols.getZeroDigit() - DEFAULT_DECIMAL_DIGITS.charAt(0);
+
+            final char[] digitChars = new char[DEFAULT_DECIMAL_DIGITS.length()];
+            for (int i = 0; i < DEFAULT_DECIMAL_DIGITS.length(); ++i) {
+                digitChars[i] = (char) (DEFAULT_DECIMAL_DIGITS.charAt(i) + zeroDelta);
+            }
+
+            return String.valueOf(digitChars);
+        }
+
+        /** Construct a new double format function.
+         * @return format function
+         */
+        public DoubleFunction<String> build() {
+            return factory.apply(this);
+        }
+    }
+
+    /** Base class for standard double formatting classes.
+     */
+    private abstract static class AbstractDoubleFormat
+        implements DoubleFunction<String>, ParsedDecimal.FormatOptions {
+
+        /** Maximum precision; 0 indicates no limit. */
+        private final int maxPrecision;
+
+        /** Minimum decimal exponent. */
+        private final int minDecimalExponent;
+
+        /** String representing positive infinity. */
+        private final String postiveInfinity;
+
+        /** String representing negative infinity. */
+        private final String negativeInfinity;
+
+        /** String representing NaN. */
+        private final String nan;
+
+        /** Flag determining if fraction placeholders should be used. */
+        private final boolean fractionPlaceholder;
+
+        /** Flag determining if signed zero strings are allowed. */
+        private final boolean signedZero;
+
+        /** String containing the digits 0-9. */
+        private final char[] digits;
+
+        /** Decimal separator character. */
+        private final char decimalSeparator;
+
+        /** Thousands grouping separator. */
+        private final char groupingSeparator;
+
+        /** Flag indicating if thousands should be grouped. */
+        private final boolean groupThousands;
+
+        /** Minus sign character. */
+        private final char minusSign;
```

Comment:
@darkma773r for no special reason, other than curiosity, I decided to focus my review using an example with negative numbers.

Found one typo, but rest of the code looks really good! I left a question somewhere in this revuiew about negative numbers I think, but here's another one:

* What should we do about using a digit for the minus sign? Is that valid in other Java classes?

e.g.

```java
        DoubleFunction<String> fmt = DoubleFormat.PLAIN.builder()
                .allowSignedZero(true)
                .minusSign('1')
                .build();
        String s = fmt.apply(-1.3);
        System.out.println(s); // gives 11.3
```

In this case, I think the user got what s/he asked for. Since `1` is the minus sign, well, we just return the value `-1.3` formatted as `11.3`. However, since it's a valid decimal, I wonder if that wouldn't cause problems to users, and problems that could be hard to troubleshoot/debug.

One option would be to throw an error for digits… but I am not sure if that's a good idea. After all, there could be other characters that would cause confusion, like using `+` for `minusSign`, or `E`. Perhaps just add a note in the docs for users? WDYT?

Reply:
Good point. I think that we can leave this as-is since the code in your example is doing exactly what was requested, even though it's not a very good idea. The JDK also does not perform any validation on this front. For example, the code below produces the same output as your example.
```java
DecimalFormatSymbols sym = new DecimalFormatSymbols();
sym.setMinusSign('1');

DecimalFormat fmt = new DecimalFormat("0.0##", sym);

System.out.println(fmt.format(-1.3)); // gives 11.3
```